### PR TITLE
Refactor buffer_period to driver-level

### PR DESCRIPTION
**Pull request by lornasong**
Discovered a bug in earlier refactor https://github.com/hashicorp/consul-terraform-sync/pull/185.
If a task is configured as disabled from the start, a template will not be
generated and template == nil. readwrite.setTemplateBufferPeriods() calls
driver.TemplateID() which throws a nil pointer panic.

Refactor setTemplateBufferPeriods():
 - Add controller.getTemplateBufferPeriods() and then set buffer info on driver.Task
 - Add driver.SetBufferPeriods() which sets buffer info from driver.Task on watcher
 - Update readwrite.Run() to call driver.SetBufferPeriods()
 - Remove the need for driver.TemplateID()

### CODE REVIEW

**Comment by lornasong** [comment]
Named this PR after the refactor since the bug can't be replicated with existing code (we don't support disabled tasks). But from a code perspective, it is a bug (unhandled pointer).

File: driver/task.go [hunk 1]
```diff
@@ -27,6 +34,7 @@ type Task struct {
 	VarFiles        []string
 	Version         string
 	UserDefinedMeta map[string]map[string]string
+	BufferPeriod    *BufferPeriod
```
Comment (eikenb) [inline driver/task.go:37]: I don't think this needs to be a pointer. For changes to the Task, it is already a pointer. I don't see any mutations of BufferPeriod in place and I don't see any cases where the nil case is needed (checking for unset doesn't seem required over 0-case) in either this PR now #185.

**Review by findkim — APPROVED** [review]
If I'm understanding correctly, the bug exists for your upcoming work for disabling tasks and not reproducible in main branch. The changes seem to work and follows the existing functionality:
* Defaults to global buffer period, overrides with task buffer period if enabled
* Disabled global + task enabled, buffer is set
* Disabled global + task, no buffer is set

File: driver/task.go (hunk 1 above)
Comment (eikenb) [reply driver/task.go:37]: I suggest this as the semantics for 0-values is much nicer (no panics) and values allow for better performance (can be allocated on the stack, so less GC).

Comment (lornasong) [reply driver/task.go:37]: Yea, there isn't any mutations of BufferPeriod. I made BufferPeriod a pointer so that the nil case would be meaningful: nil = buffer_period is disabled.

The alternative would be instead to have an enabled field on the BufferPeriod. This was a little less intuitive to me from the backend since disabled BufferPeriod should not have mix/max values. However, this would be more in line with the configuration structure 🤔 .

What do you think? You mention a lot of other positives for not having a pointer. Is your preference still for not having a pointer? Happy to change, just want to double-check.

**Comment by lornasong** [comment]
thanks @findkim! yes, you are understanding correctly that the bug is not reproducible in the main branch and exists for upcoming work. thanks for confirming the different scenarios!

File: driver/task.go (hunk 1 above)
Comment (eikenb) [reply driver/task.go:37]: If nil is used as an unset/disabled value where the 0 value won't work, then it makes sense. I probably missed or misinterpreted where that was done. Using an enabled field is an option but I don't think it is a slam dunk improvement as it has downsides as well.

TLDR; it is being used to represent 'disabled' and no change is necessary.

**Review by eikenb — APPROVED** [review]

File: driver/task.go (hunk 1 above)
Comment (lornasong) [reply driver/task.go:37]: Gotcha! Thanks for your input and helpful to hear your train of thought. Let me add a comment to clarify that the nil value is 'disabled' since it's not not very apparent.
